### PR TITLE
added modify[En,De]codingHeaders(), and getTcpInfo() methods

### PR DESCRIPTION
**Pull request by forkjoseph**
Commit Message:
While there're modifyEncodingData() and modifyDecodingData() features,
a feature for header modificaiton is missing. This feature can be useful
for HTTP filters that would like to redirect destination of a message
based on load, latency, bw condition, etc.
Linux > 2.4 provides TCP statistics via TCP_INFO socket option. The
stats can be used by external HTTP filters to take proper action based
on various metrics such as minRTT, congestion window, etc. We propose
getTcpInfo() in connection implementation as an experimental mode,
turned on/off by --copt="-D ENABLE_TCP_INFO=1" bazel argument.

Additional Description:
Risk Level: Low
Testing: along with "//test/", we have tested on AKS environment.
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: include/envoy/http/filter.h [hunk 1]
```diff
@@ -240,6 +240,12 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual void modifyDecodingBuffer(std::function<void(Buffer::Instance&)> callback) PURE;
 
+  /**
+   * Allows modifying the decoding header. May only be called before any header has been continued
+   * past the calling filter.
+   */
+  virtual void modifyDecodingHeaders(std::function<void(Http::HeaderMap&)> callback) PURE;
```
Comment (mattklein123) [inline include/envoy/http/filter.h:247]: These functions aren't actually needed. You can store and modify the headers when you get decode/encode headers.

File: include/envoy/network/connection.h [hunk 2]
```diff
@@ -139,6 +139,19 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
...
+  {
+    uint32_t cwnd;      /* congestion window (packets) */
+    uint32_t in_flight; /* packets "in flight" */
+    uint32_t min_rtt;   /* minimum RTT in microsecond */
+    uint32_t rtt;       /* RTT in microsecond */
+    uint64_t delivery_rate;  /* bytes per second */
+  };
+
+  virtual absl::optional<TCPInfo> getTCPInfo() const PURE;
+#endif
```
Comment (mattklein123) [inline include/envoy/network/connection.h:153]: I like the idea of exposing this, but can we please do this in a dedicated PR? I would suggest defining a common TCPInfo struct as you have done, and then hiding it behind a platform API in which non-supported platforms can just return nullopt. We shouldn't expose the ifdefs to most of the code.

Comment (forkjoseph) [reply include/envoy/network/connection.h:153]: Thanks for the comment Matt. You're right, we shouldn't expose ifdefs to most of the code. 
Let me work on it this week and submit another PR for TCPInfo :)

File: include/envoy/http/filter.h (hunk 1 above)
Comment (forkjoseph) [reply include/envoy/http/filter.h:247]: Hmm, maybe I got it wrong. But, from what I experienced, a filter has an access to the request/response header map only when encode/decodeHeader is invoked. If the filter wants to append custom header after inspecting the payload (en,decodeData function), it doesn't have a way to modify the header, unless modifyDecodingHeaders function can handle? Please feel free to correct me if I got it wrong.

Comment (mattklein123) [reply include/envoy/http/filter.h:247]: You can store a reference to the header map and modify it later before you continue iteration.
